### ISSUE TITLE
Custom Training and Inference 

### ISSUE THREAD

**Issue by abhigoku10** [edwardzhou130/PolarSeg#2]
@YangZhang4065 @edwardzhou130 Thanks for open sourcing the work , its of great work . I am having few queries 
Q1. You have tested on the semattic kitti dataset , how will the output differ when we reduce the point cloud density of the scene to 80%-90% less 
Q2.The inference you tested on which sys configuration
Q3 Can we train the polar seg for less number of class like 8 if so what al modification i have to do it for the code 
Thanks in advance

**Comment by YangZhang4065** [comment]
Hi @abhigoku10 ,

Thank you for your interest!
**Q1**: We haven't tried to reduce the point density. You are welcome to try it out!
You can also try a smaller quantization size to gain speed/ FPS as our framework is robust to different feature map sizes as well. We used 480X360 in our paper but we observe 320X210 (2.5 times smaller feature map) only decrease SemanticKITTI validation mIoU by 2.2% to 52.7%.

**Q2**: We benchmarked ours & baselines' speed, FPS, etc. on one single TITAN Xp and Xeon(R) Silver 4110 CPU @ 2.10GHz. CPU was not the bottleneck and GPU was fully utilized.

**Q3**: Sure you can use your own dataset. You will need to modify
```python
my_BEV_model=BEV_Unet(n_class=len(unique_label), n_height = compression_model, input_batch_norm = True, dropout = 0.5, circular_padding = circular_padding)
```
to
```python
my_BEV_model=BEV_Unet(n_class=your_training_data_num, n_height = compression_model, input_batch_norm = True, dropout = 0.5, circular_padding = circular_padding)
```
in training code.
You will also have to come up with your own data generator to replace
```python
    train_pt_dataset = SemKITTI(data_path + '/sequences/', imageset = 'train', return_ref = True)
    val_pt_dataset = SemKITTI(data_path + '/sequences/', imageset = 'val', return_ref = True)
```
. Your customized points generators should yield xyz (matrix of shape [n,3]), label (matrix of shape [n,]), feature (matrix of shape [n,m]. Optional. Could be reflection or RGB.) upon iterated. The xyz should be centered at the sensor. Refer to `dataset.py` for more info.
Note that you need to make sure your labels range from `0`-`(L-1)`, where `L` is the number of the meaningful classes you want to train. Any background classes that you want to ignore should be mapped to 255. You can do that by modifying `semantic-kitti.yaml`.

**Comment by polars05** [comment]
@YangZhang4065, if say I wanted to ignore bicycles, does it means I should modify the bicycle entry in the learning map in ```semantic-kitti.yaml``` from ```11: 2``` to ```11: 255```?

https://github.com/edwardzhou130/PolarSeg/blob/172c6505183fee50da9a4563f5170804bf214f91/semantic-kitti.yaml#L109-L113

And in that case, should I also set ```ignore_index``` and and ```ignore_label``` to ```255```in the two sections of code below?:
https://github.com/edwardzhou130/PolarSeg/blob/172c6505183fee50da9a4563f5170804bf214f91/train.py#L77

https://github.com/edwardzhou130/PolarSeg/blob/172c6505183fee50da9a4563f5170804bf214f91/train.py#L83-L84

**Comment by YangZhang4065** [comment]
Hi @polars05,

**TL; DR**: You only need to change bicycle entry in the learning map in `semantic-kitti.yaml` from `11: 2` to `11: 0` and nothing else.

Explain:
The `ignore_label`  in 
https://github.com/edwardzhou130/PolarSeg/blob/172c6505183fee50da9a4563f5170804bf214f91/train.py#L83-L84
actually refers to what background label we used to fill the empty space after quantization when we are doing quantization in the data loader.  We choose to fill all vacant space with 0 (`unlabeled` in SmeanticKITTI). You should not change `ignore_label` to 255 yet as 255 is a label that is used to represent moving-motorcyclist in SemanticKITTI. You want the output of your data loader to be consistent with Semantic KITTI.

All those background `0`s are later mapped to 255 in training code in the following code, along with other `0`s in the original dataset.

https://github.com/edwardzhou130/PolarSeg/blob/172c6505183fee50da9a4563f5170804bf214f91/train.py#L41-L45

**Comment by abhigoku10** [comment]
@YangZhang4065  thanks for the response , i have following queries 
Q1. From the paper you have compared spherical , cartesian and polar projection with other backbone like squeeze seg , resnet and unet . I want to try it with other architecture like yolact[resnet+fpn], yolov3[darknet] can i do so to check the accuracy and performance which part of the code should i modify 
Q2. How to compare viz the polar bev vs cartesian bev , can i save the images since i want to compare few corner cases of different scenes
Q3. Can Polar net used for 3d object detection feature also 
Q4. Can polar net be modified for instance semanttic segmentation 
Q5. My custom dataset has less numbers of point density when compared to semanttic kitti should i train the model from sctrach or use the pretrained weights 

thanks a lot in advance

**Comment by edwardzhou130** [comment]
Hi @abhigoku10,

**Q1.** You can replace the backbone network in here:
https://github.com/edwardzhou130/PolarSeg/blob/172c6505183fee50da9a4563f5170804bf214f91/network/BEV_Unet.py#L16
Theoretically, you can use any 2D semantic segmentation backbone you like as long as it has the same input (**BxFxHxW**, where F is the encoding feature size in our polar BEV representation) and output size (**BxCxHxWxD**). However, as discussed in section 3.4 in our paper, you need to change the convolution layer to ring convolution to fully benefit from the polar BEV representation.

**Q2.**  You can use other point cloud libraries like [Open3D](https://github.com/intel-isl/Open3D) to visualize the result. The color map of each class is provided in `semantic-kitti.yaml`.

**Q3/4.** We developed PolarNet specifically for the 3D LiDAR point cloud semantic segmentation task. 

**Q5.** Our pretrained weights are highly optimized for the SemanticKITTI dataset. For example, the 3D space range and the class distribution, etc. I recommend training the model from scratch if you are using your custom dataset.
